### PR TITLE
Adds global filters as well as custom handling for duplicates and misses

### LINKED ISSUE #108: Improve fluent API for more flexibility
Now | Next
-- | --
Map<Event>().As().When() | Map<Event>().As()
Map<Event>().AsCreateOf().Using() | Map<Event>().AsCreateOf().Using().ThrowingOnDuplicates()
Map<Event>().AsCreateIfDoesNotExistsOf().Using() | Map<Event>().AsCreateOf().Using().IgnoringDuplicates()   
|| Map<Event>().AsUpdateOf().Using().CreatingIfMissing()
  | Map<Event>().AsCreateOf().Using().HandlingDuplicatesUsing(x   => x.)
Map<Event>().AsCreateOrUpdateOf().Using() | Map<Event>().AsCreateOf().Using().OverwritingDuplicates()
Map<Event>().AsUpdateOf().Using() | Map<Event>().AsUpdateOf().Using().ThrowingIfMissing()
Map<Event>().AsUpdateIfExists().Using() | Map<Event>().AsUpdateOf().Using().IgnoringMisses()
  | Map<Event>().AsUpdateOf().Using().HandlingMissesUsing(x   => x)
Map<Event>().AsDeleteOf().Using() | Map<Event>().AsDeleteOf().Using().ThrowingIfMissing()
Map<Event>().AsDeleteIfExists().Using() | Map<Event>().AsDeleteOf().Using().IgnoringMisses()
  | Map<Event>().AsDeleteOf().Using().HandlingMissesUsing(x   => x)

### PR DESCRIPTION
**Pull request by dennisdoomen**
Prepares for #110 
Closes #108 and #101 

*Release notes*
* {New} Added support for applying filters on the `EventMapBuilder` level using `Where` - #111
* {New} `IEventMap.Handle` will now return a boolean to indicate whether or not the event was configured to be handled (accounting for the filters applied) - #111
* {Breaking} The contract for a projector that is needed to build an `IEventMap` have changed. The methods `ProjectionModificationHandler`, `HandleProjectionDeletionsAs` and such on `EventMapBuilder` are now passed to the `Build` method as an instance of `ProjectionMap` - #111

### CODE REVIEW

**Comment by IharBury** [comment]
Why is `.vs/LiquidProjections/v15/sqlite3/storage.ide` included?

File: Src/LiquidProjections/EventMapBuilder.cs [hunk 1]
```diff
@@ -12,98 +11,67 @@ namespace LiquidProjections
...
-            return new EventMappingBuilder<TEvent>(this);
+            return new Action<TEvent>(this, () => projector);
         }
 
         /// <summary>
         /// Builds the resulting event map. Can only be called once.
         /// No changes can be made after the event map has been built.
         /// </summary>
-        public IEventMap<TContext> Build()
+        public IEventMap<TContext> Build(ProjectorMap<TContext> projector)
```
Comment (IharBury) [inline Src/LiquidProjections/EventMapBuilder.cs]: The argument is not documented

File: Src/LiquidProjections/EventMapBuilder.cs [hunk 2]
```diff
@@ -12,98 +11,67 @@ namespace LiquidProjections
...
-            customHandler = handler;
-        }
-
-        internal void Add<TEvent>(Func<TEvent, TContext, Task> action)
-        {
-            if (action == null)
+            if (projector.Custom == null)
             {
-                throw new ArgumentNullException(nameof(action));
+                throw new ArgumentNullException(nameof(projector.Custom));
```
Comment (IharBury) [inline Src/LiquidProjections/EventMapBuilder.cs]: "Custom" is not a name of an argument of the method

File: Src/LiquidProjections/EventMapBuilder.cs [hunk 3]
```diff
@@ -150,1125 +118,272 @@ internal void Add(Func<TEvent, TContext, Task> action)
...
 
-                private readonly ProjectionEventMappingBuilder<TEvent> eventMappingBuilder;
-                private readonly Func<TEvent, TKey> getKey;
+                    return this;
+                }
 
-                public CreateIfDoesNotExistEventActionBuilder(
-                    ProjectionEventMappingBuilder<TEvent> eventMappingBuilder,
-                    Func<TEvent, TKey> getKey)
+                public ICreateAction<TEvent, TProjection, TContext> IgnoringDuplicates()
```
Comment (IharBury) [inline Src/LiquidProjections/EventMapBuilder.cs]: I don't like that you can combine several calls like `.IgnoringDuplicates().OverwritingDuplicates()` in one registration, in which case the last one wins. This also affects the old style API like `.AsCreateIfDoesNotExistOf().OverwritingDuplicates()`.

File: Src/LiquidProjections/EventMapBuilder.cs [hunk 4]
```diff
@@ -69,6 +80,11 @@ public Action(EventMapBuilder<TContext> parent, Func<ProjectorMap<TContext>> get
             {
                 this.parent = parent;
                 this.getProjector = getProjector;
+
+                foreach (Func<object,TContext,Task<bool>> predicate in parent.predicates)
```
Comment (IharBury) [inline Src/LiquidProjections/EventMapBuilder.cs]: This way the map-level filters only apply to event handlers registered after the filters are registered.

File: Src/LiquidProjections/EventMap.cs [hunk 5]
```diff
@@ -26,7 +24,7 @@ internal void Add<TEvent>(Func<TEvent, TContext, Task> action)
         /// <summary>
         /// Handles <paramref name="anEvent"/> asynchronously using context <paramref name="context"/>.
         /// </summary>
-        public async Task Handle(object anEvent, TContext context)
+        public async Task<bool> Handle(object anEvent, TContext context)
```
[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/LiquidProjections/EventMap.cs]: This change is non-breaking. You don't have to use it.

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply Src/LiquidProjections/EventMap.cs]: And the docs are on the interface.

File: Src/LiquidProjections/EventMapBuilder.cs (hunk 3 above)
Comment (dennisdoomen) [reply Src/LiquidProjections/EventMapBuilder.cs]: I see that, but it is pretty self-explanatory if you do. Don't feel like fixing that.

File: Src/LiquidProjections/EventMap.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (IharBury) [reply Src/LiquidProjections/EventMap.cs]: https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/breaking-change-rules.md#property-field-parameter-and-return-values

**Pull request merged**